docs/concepts/streaming-actions.rst

```diff
@@ -0,0 +1,91 @@
+=================
+Streaming Actions
+=================
+
+.. _streaming_actions:
+
+Actions can be implemented as streaming results. This enables a lower time-to-first-token and a more interactive
+interface in the case of AI applications or streaming in of metrics in a model-training application. Broadly,
+this is a tool to enable quicker user interaction in longer running actions that require user focus.
+
+Streaming actions can be implemented as a class or a function, just like actions. However, they have a few additional rules:
+
+#. They give intermediate results to the user as they are generated in the form of a generator
+#. They are responsible for determining the relationship between the intermediate results and the final result
+#. They have a ``return`` statement that handles result collection and state update
+
+If you're not familiar with ``return`` statements in generators, you can read about them `here <https://www.python.org/dev/peps/pep-0380/>`_.
+The high-level idea is that the ``return`` statement is used to raise a ``StopIteration`` exception with a value that is returned to the caller.
+The framework uses this value to update the state of the action and to collect the final result.
+
+A class-based streaming action might look like this:
+
+.. code-block:: python
+
+    from burr.core.action import StreamingAction
+
+    class StreamingChatCall(StreamingAction):
+        def stream_run(self, state: State, **run_kwargs) -> Generator[dict, None, dict]:
+            client = openai.Client()
+            response = client.chat.completions.create(
+                model='gpt-3.5-turbo',
+                messages=[{
+                    'role': 'user',
+                    'content': state["prompt"]
+                }],
+                temperature=0,
+                stream=True,
+            )
+            buffer = []
+            for chunk in response:
+                delta = chunk.choices[0].delta.content
+                buffer.append(delta)
+                yield {'response': delta}
+            full_response = ''.join(buffer)
+            return {'response': full_response}
```

Comment:
again, why no state return? It says it writes to it below, but that isn't true. Otherwise this should be an error condition, no?

Reply:
Oooooooh I see -- this is the Object oriented approach.